spec/abci++/abci++_properties_001_draft.md

```diff
@@ -0,0 +1,675 @@
+---
+order: 1
+title: New Methods
+---
+
+# New Methods
+
+## Description
+
+### PrepareProposal
+
+#### Parameters and Types
+
+>**TODO**: Hyperlinks for ConsensusParams, LastCommitInfo, Evidence, Event, and ValidatorUpdate are
+broken because they are defined in abci.md (and not copied over for the moment).
+
+>**TODO**: don't forget to extend the Header structure with "bool problem" when including the parts
+common with the existing ABCI spec.
+
+* **Request**:
+
+    | Name                    | Type                                        | Description                                                                                                | Field Number |
+    |-------------------------|---------------------------------------------|------------------------------------------------------------------------------------------------------------|--------------|
+    | hash                    | bytes                                       | The hash of the block to propose. This can be derived from the block header.                               | 1            |
+    | header                  | [Header](../core/data_structures.md#header) | The header of the block to propose.                                                                        | 2            |
+    | last_commit_info        | [LastCommitInfo](#lastcommitinfo)           | Info about the last commit, including the round, the validator list, and which ones signed the last block. | 3            |
+    | byzantine_validators    | repeated [Evidence](#evidence)              | List of evidence of validators that acted maliciously.                                                     | 4            |
+    | tx                      | repeated bytes                              | Preliminary list of transactions that have been picked as part of the block to propose.                    | 5            |
```

Comment:
Oh, I see now. There are two possible actions that can be taken by an application. The application can mark the Tx as 'remove', in which case the Tx will be removed from the mempool. Alternatively, the application can _remove_ the application from the Tx list, in which case it stays in the mempool but is not in the proposal. 

I think we should stick with one format for modification of the list of Tx; namely, I think that if an application wants to alter a Tx, it should have to return the annotated Tx with the enum value relaying what behavior it wants. If you want it to not appear in _this specific block_, then marking it as 'UNPROPOSE' or 'REDACT' or similar seems a bit clearer. This is somewhat of a minor point either way. This would also allow the list of Tx returned to tendermint to be just a list of transformations or modifications. Absence from the list of modifications means 'no changes to this Tx' instead of 'do not propose at this time'.

Reply:
Ok. let me see if I understood what you are proposing. The changes would be:
* Delete the "UNMODIFIED" value from the `TxAction` enum.
* Added a "DELAYED" or "UNPROPOSE" value to the `TxAction` enum
If a transaction is to appear unmodified, we simply remove it from the list. If it is to be removed from the proposal but kept in the mempool, we mark it as "UNPROPOSE".
If my understanding is correct, I see an issue with this solution. Since we are to support transaction reordering, I don't clearly see how we could combine reordered, modified, and unmodified transactions, since we would remove the unmodified ones from the list.

Reply:
That's a great point and as a result I think that my suggestion is incompatible with the requirement of allowing Tx to be reordered.